File: Data Structures & Algorithms/search-2d-matrix/submission-6.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int l=0;
        int r=matrix.size()-1;
        while(l<r){
            int m = l+(r-l)/2;
            if(target > matrix[m][matrix[0].size()-1]){
                l=m+1;
            }
            else if(target < matrix[m][0]){
                r=m-1;
            }
            else{
                break;
            }
        }
        int l2 =0;
        int row = (l+r)/2;
        int r2 = matrix[row].size();
        while(l2<=r2){
            int m = l2+(r2-l2)/2;
            if(matrix[row][m] < target){
                l2 = m+1;
            }
            else if(matrix[row][m] > target){
                r2 = m-1;
            }
            else{
                return true;
            }
        }
        return false;

    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-6.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        row = len(grid)
        col = len(grid[0])
        islands = 0

        def bfs(r, c):
            q = collections.deque()
            q.append((r,c))
            while q:
                a,b = q.popleft()
                directions = [[1,0], [0,1], [-1,0], [0,-1]]
                
                for dr, dc in directions:
                    if ((a+dr) in range(row) and (b+dc) in range(col) and grid[a+dr][b+dc]=="1"):
                        grid[a+dr][b+dc]="0"
                        q.append((a+dr, b+dc))


        for r in range(row):
            for c in range(col):
                if grid[r][c]=="1":
                    bfs(r,c)
                    islands+=1
        return islands

